cupyx/scipy/interpolate/_polyint.py

```diff
@@ -0,0 +1,323 @@
+                                           self._y_extra_shape[:-self._y_axis])
+            raise ValueError("Data must be of shape %s" % ok_shape)
+        return yi.reshape((yi.shape[0], -1))
+
+    def _set_yi(self, yi, xi=None, axis=None):
+        if axis is None:
+            axis = self._y_axis
+        if axis is None:
+            raise ValueError("no interpolation axis specified")
+
+        shape = yi.shape
+        if shape == ():
+            shape = (1,)
+        if xi is not None and shape[axis] != len(xi):
+            raise ValueError("x and y arrays must be equal in length along "
+                             "interpolation axis.")
+
+        self._y_axis = (axis % yi.ndim)
+        self._y_extra_shape = yi.shape[:self._y_axis]+yi.shape[self._y_axis+1:]
+        self.dtype = None
+        self._set_dtype(yi.dtype)
+
+    def _set_dtype(self, dtype, union=False):
+        if cupy.issubdtype(dtype, cupy.complexfloating) \
+                or cupy.issubdtype(self.dtype, cupy.complexfloating):
+            self.dtype = cupy.complex_
+        else:
+            if not union or self.dtype != cupy.complex_:
+                self.dtype = cupy.float_
+
+
+class BarycentricInterpolator(_Interpolator1D):
+    """The interpolating polynomial for a set of points.
+
+    Constructs a polynomial that passes through a given set of points.
+    Allows evaluation of the polynomial, efficient changing of the y
+    values to be interpolated, and updating by adding more x values.
+    For reasons of numerical stability, this function does not compute
+    the coefficients of the polynomial.
+
+    The value `yi` need to be provided before the function is
+    evaluated, but none of the preprocessing depends on them,
+    so rapid updates are possible.
+
+    Parameters
+    ----------
+    xi : array-like
+        1-D array of x-coordinates of the points the polynomial should
+        pass through
+    yi : array-lie, optional
+        The y-coordinates of the points the polynomial should pass through.
+        If None, the y values will be supplied later via the `set_y` method
+    axis : int, optional
+        Axis in the yi array corresponding to the x-coordinate values
+
+    See Also
+    --------
+    scipy.interpolate.BarycentricInterpolator
+
+    """
+
+    def __init__(self, xi, yi=None, axis=0):
+        _Interpolator1D.__init__(self, xi, yi, axis)
+
+        self.xi = cupy.asfarray(xi)
+        self.set_yi(yi)
+        self.n = len(self.xi)
+
+        self._inv_capacity = 4.0 / (cupy.max(self.xi) - cupy.min(self.xi))
+        permute = cupy.random.permutation(self.n)
+        inv_permute = cupy.zeros(self.n, dtype=cupy.int32)
+        inv_permute[permute] = cupy.arange(self.n)
+
+        self.wi = cupy.zeros(self.n)
+        for i in range(self.n):
+            dist = self._inv_capacity * (self.xi[i] - self.xi[permute])
+            dist[inv_permute[i]] = 1.0
+            self.wi[i] = 1.0 / cupy.prod(dist)
```

Comment:
If possible, we'd like to perform this operation without a for loop, but (batched) ndarray operations only.